# Project Title
Comparison of Logistic Regression and SVM Method

## Members

1. First member: name and e-mail adress
Zeynep Göger and gogerz@itu.edu.tr
2. Second member: name and e-mail adress

## Description of the project
In this project,  we compare two methods that are logistic regression method and SVM method 
by using 1984 United Stated Congressional Voting Records in R language. We analyse that which
method is better to predict the congressman belongs to which party according to first column
of the data to the rest of them.


### The methods to be used
There are two different methods that are logistic regression method and SVM method.
Logistic regression is a type of a prediction method that is used if the data has binary 
dependent variables and this method determines the relationship between the set of dependent
variables and the independent set.
    
Support Vector Machines method is another prediction method that is used if the data is
binary too. However, in this method two kinds of accumulation of dependent variables
are determined and divided down the middle of the distance that is called as margin.    

Explain.

### The data

The data is 1984 United Stated Congressional Voting Records that has 435 rows and 17 columns. This data consists of 435 congressmen that are 267 democrats and 168 republicans.
In this data has binary dependent variables such as yes or no and two class of congressmen
that are republican and democrat. There are 16 numbers of attributes that are categorical.
    
## Code

As proof of work, you must run this notebook.  Upload an HTML output of this notebook on your github account.



In [40]:
X<- read.csv("https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data", header=FALSE)

In [41]:
head(X)

V1         V2 V3 V4 V5 V6 V7 V8 V9 V10 V11 V12 V13 V14 V15 V16 V17
1 republican n  y  n  y  y  y  n  n  n   y   ?   y   y   y   n   y  
2 republican n  y  n  y  y  y  n  n  n   n   n   y   y   y   n   ?  
3 democrat   ?  y  y  ?  y  y  n  n  n   n   y   n   y   y   n   n  
4 democrat   n  y  y  n  ?  y  n  n  n   n   y   n   y   n   n   y  
5 democrat   y  y  y  n  y  y  n  n  n   n   y   ?   y   y   y   y  
6 democrat   n  y  y  n  y  y  n  n  n   n   n   n   y   y   y   y

## 1.Logistic Regression
### a.Setting up Logistic Regression Model
Logistic regression is a type of a prediction method that is used if the data has binary 
dependent variables and this method determines the relationship between the set of dependent
variables and the independent set. Odds is the ratio of the probability of occurrence with probability of event that does not occur.
$$odds= p/(1-p) $$
$$logit p = log(odds) = log(p/(1-p)) = \beta_0 + \beta_1x_1 +...+  \beta_k x_k$$
This formula resembles a linear regression but logistic regression gets a best fitting by using maximum likelihood function that maximizes the probability. $$p=\frac{e^{\beta_0 + \beta_1x_1 +...+  \beta_k x_k}}{1+e^{\beta_0 + \beta_1x_1 +...+  \beta_k x_k}}$$ where $\beta_0$ is constant,$\beta_1,...,\beta_k$ are the coeffients of predictor variables, $e$ is a natural logarithm and $p$ is the probability of a specific case.

Also we can write this equation as the following
$$ e(logit p)= e^{(\beta_0) + (\beta_1x_1) +...+  (\beta_k x_k)}$$  then
$$o= p/(1-p)=e^{(\beta_0)}e^{(\beta_1x_1)}e^{(\beta_2x_2)}....e^{(\beta_kx_k)}$$
where $\beta_k'$s are constant coefficients that determines the changes of logistic regression model when $x_i$ is 
added. Also, adding or subtracting a unit alters the odds with constant amount.
### b.The Main Goals of Logistictic Regression Model
There are two primary purposes of logistic regression models. First one of these purposes is to predict of group membership.This model uses the odds ratio;therefore, the conclusion of using this model has the the same form with the odds ratio. The second aim is suplying informations about connections and strongness between the variables. 
### c.Setting up a Model in R
Initially, we divide the data to constitute our model and we divide 30% of the data randomly. To do this we use this code. 

In [42]:
sample(1:nrow(X), 0.3*nrow(X))
test <-sample(1:nrow(X), 0.3*nrow(X))

[1] 312 239  80 199 317 306 155 425 235 413 274 295 410 138 242  38  68 298
 [19] 115 422 412 299  22 248  45  56 258 236 218 391  59 210  72 195 112 251
 [37]  13 131 101   4 227 334 186 389 283  57 270 393 324 340 362  89 273 106
 [55] 121   8 116  24 423 148 431 136  66  12 276 253 127  71 311 198 123 245
 [73]  79  35   5 173 243 144 414 104 214 313 261 184 263 409 326 296 203 359
 [91]  14 375  78  23  67 175 119  20 354 141 361  85 380 286 172 322  88 237
[109] 384 339  28 429  50 241  73 302 405  18 371 132 407 328   3 182 200  10
[127] 187 194 134 350

In [43]:
length(test)

[1] 130

130 members of this data is selected randomly by using this code in R but these numbers are not useful for testing and forming our model. Therefore, we convert them to a matrix by using the code below.

In [44]:
X[test,]

V1         V2  V3  V4  V5  V6  V7  V8  V9  V10 V11 V12 V13 V14 V15 V16 V17
25  democrat   y   n   y   n   n   n   y   y   y   n   n   n   n   n   y   ?  
370 republican n   y   n   y   y   y   n   n   n   n   n   n   y   y   n   y  
69  democrat   y   ?   y   n   n   n   y   y   y   n   n   n   y   n   y   y  
290 democrat   y   y   y   n   y   y   n   y   n   y   y   n   y   y   n   ?  
154 democrat   n   n   y   n   n   y   y   y   y   y   y   n   y   y   n   y  
171 democrat   y   n   y   n   n   n   ?   y   y   ?   n   n   n   n   y   ?  
420 democrat   y   y   y   n   n   n   y   y   y   n   n   n   n   n   n   y  
95  democrat   y   n   y   n   y   y   n   n   n   n   n   n   n   n   n   y  
130 democrat   ?   ?   y   n   n   n   y   y   ?   n   ?   ?   ?   ?   ?   ?  
365 republican y   y   n   y   y   y   n   n   n   n   y   n   y   y   n   y  
244 democrat   n   y   y   n   n   n   y   y   ?   y   n   n   y   n   y   y  
230 republican n   y   y   y   y   y   y   n   y   y   n   y   y   y   n   y  
205 republican n   n   y   y   y   y   y   n   n   n   n   y   y   y   n   y  
304 republican n   n   n   y   y   y   n   n   n   n   n   y   y   y   n   n  
53  democrat   y   y   y   n   n   n   y   y   y   n   n   n   n   n   ?   ?  
183 democrat   y   ?   y   n   n   n   y   y   y   n   n   n   n   n   y   y  
286 democrat   n   n   y   n   n   y   y   y   y   y   y   n   n   n   y   y  
65  democrat   y   y   y   n   n   n   y   y   y   n   y   ?   n   n   n   y  
275 republican y   n   n   y   y   n   y   n   n   y   n   n   n   y   y   y  
265 democrat   y   n   y   n   n   n   y   y   y   n   n   n   n   n   y   ?  
115 democrat   y   ?   y   n   n   n   y   y   y   n   n   n   n   n   y   ?  
225 republican n   y   n   y   y   y   n   n   n   y   n   ?   y   y   n   n  
407 democrat   y   n   y   n   y   y   n   n   y   y   n   n   y   y   n   y  
260 democrat   y   n   y   n   n   n   y   y   y   n   n   n   n   n   n   y  
332 democrat   y   ?   y   n   n   n   y   y   y   n   n   n   n   n   y   ?  
28  democrat   y   y   y   n   n   n   y   y   y   n   y   n   n   n   y   y  
125 democrat   n   ?   y   n   n   n   y   y   y   n   n   n   n   n   y   y  
151 republican y   y   n   y   y   y   n   n   n   y   n   y   y   y   n   y  
43  democrat   y   n   y   n   n   n   y   y   y   n   n   n   n   n   n   y  
76  democrat   n   y   y   y   y   y   n   n   n   y   y   n   y   y   n   n  
... ...        ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... ...
381 democrat   y   y   y   n   n   y   ?   y   y   n   y   n   y   n   y   y  
259 democrat   n   n   y   n   n   n   y   y   y   n   y   n   n   n   y   y  
6   democrat   n   y   y   n   y   y   n   n   n   n   n   n   y   y   y   y  
44  democrat   y   n   y   n   n   n   y   y   y   n   n   n   n   n   y   y  
424 democrat   n   y   y   n   n   y   y   y   y   n   y   n   n   y   y   y  
93  democrat   y   y   y   n   n   n   y   y   n   y   y   n   n   ?   y   y  
374 democrat   n   y   y   ?   y   y   n   y   n   y   ?   n   y   y   ?   y  
140 democrat   y   n   y   n   n   y   y   y   y   n   n   n   n   n   y   y  
180 democrat   y   n   y   n   n   n   y   y   y   y   n   n   n   n   y   y  
128 democrat   n   ?   y   n   n   n   y   y   y   y   y   n   n   y   y   y  
124 republican y   ?   n   y   y   y   n   y   n   n   n   y   y   y   n   y  
98  democrat   y   n   n   n   y   y   y   n   n   y   y   n   n   y   n   y  
32  democrat   y   y   y   n   n   n   y   y   y   n   y   n   n   n   y   ?  
352 republican n   n   y   y   y   y   y   y   y   y   n   y   y   y   y   y  
261 democrat   y   n   y   n   n   n   y   y   y   y   n   n   n   n   n   y  
431 republican n   n   y   y   y   y   n   n   y   y   n   y   y   y   n   y  
343 democrat   n   y   y   n   y   ?   y   n   n   y   y   n   y   n   y   y  
77  democrat   n   y   y   n   y   y   n   n   n   y   y   y   y   y   n   ?  
295 democrat   n   y   y   n   n   y   y   y   n   y  

In [45]:
nrow(X[test,])
modeldata<-X[test,]

[1] 130

In [46]:
X[-test,]

V1         V2  V3  V4  V5  V6  V7  V8  V9  V10 V11 V12 V13 V14 V15 V16 V17
1   republican n   y   n   y   y   y   n   n   n   y   ?   y   y   y   n   y  
2   republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   ?  
3   democrat   ?   y   y   ?   y   y   n   n   n   n   y   n   y   y   n   n  
5   democrat   y   y   y   n   y   y   n   n   n   n   y   ?   y   y   y   y  
9   republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   y  
10  democrat   y   y   y   n   n   n   y   y   y   n   n   n   n   n   ?   ?  
12  republican n   y   n   y   y   y   n   n   n   n   y   ?   y   y   ?   ?  
13  democrat   n   y   y   n   n   n   y   y   y   n   n   n   y   n   ?   ?  
14  democrat   y   y   y   n   n   y   y   y   ?   y   y   ?   n   n   y   ?  
15  republican n   y   n   y   y   y   n   n   n   n   n   y   ?   ?   n   ?  
17  democrat   y   n   y   n   n   y   n   y   ?   y   y   y   ?   n   n   y  
19  republican n   y   n   y   y   y   n   n   n   n   n   ?   y   y   n   n  
20  democrat   y   y   y   n   n   n   y   y   y   n   y   n   n   n   y   y  
21  democrat   y   y   y   n   n   ?   y   y   n   n   y   n   n   n   y   y  
23  democrat   y   ?   y   n   n   n   y   y   y   n   n   ?   n   n   y   y  
24  democrat   y   y   y   n   n   n   y   y   y   n   n   n   n   n   y   y  
26  democrat   y   n   y   n   n   n   y   y   y   y   n   n   n   n   y   y  
27  democrat   y   n   y   n   n   n   y   y   y   n   y   n   n   n   y   y  
29  republican y   n   n   y   y   n   y   y   y   n   n   y   y   y   n   y  
30  democrat   y   y   y   n   n   n   y   y   y   n   y   n   n   n   y   y  
31  republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   n  
33  democrat   y   y   y   n   n   n   y   y   y   y   n   n   y   n   y   y  
34  republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   y  
35  democrat   y   y   y   n   n   n   y   y   y   n   n   n   n   n   y   y  
36  republican n   y   n   y   y   y   n   n   n   n   n   y   y   y   n   n  
37  republican y   ?   n   y   y   y   n   n   n   y   n   y   ?   y   n   y  
39  republican n   y   n   y   y   y   n   n   n   y   n   y   y   y   n   n  
40  democrat   y   n   y   n   n   n   y   y   y   y   y   n   y   n   y   y  
41  democrat   y   y   y   n   n   n   y   y   y   n   ?   n   n   n   n   ?  
42  democrat   y   y   y   n   n   n   y   y   y   n   n   n   n   n   y   ?  
... ...        ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... ...
397 democrat   y   y   y   n   y   y   n   y   n   n   y   n   y   n   y   y  
398 democrat   y   y   n   n   y   ?   n   n   n   n   y   n   y   y   n   y  
399 democrat   n   y   y   n   y   y   n   y   n   n   n   n   n   n   n   y  
400 republican n   y   n   y   ?   y   n   n   n   y   n   y   y   y   n   n  
401 republican n   y   n   y   y   y   n   ?   n   n   ?   ?   ?   y   n   ?  
402 republican n   y   n   y   y   y   n   n   n   y   y   y   y   y   n   n  
404 republican n   y   n   y   y   y   n   n   n   y   n   y   ?   y   n   n  
405 republican y   y   n   y   y   y   n   n   n   y   n   y   y   y   n   y  
406 republican n   n   n   y   y   y   n   n   n   n   n   y   y   y   n   y  
408 democrat   n   n   n   y   y   y   n   n   n   n   y   y   y   y   n   n  
409 democrat   y   n   y   n   n   y   y   y   y   n   n   y   ?   y   y   y  
410 republican n   n   n   y   y   y   n   n   n   n   n   y   y   y   n   n  
411 republican n   n   n   y   y   y   n   n   n   n   y   y   y   y   n   y  
412 democrat   y   n   y   n   n   y   y   y   y   y   y   n   n   n   n   y  
413 republican n   n   n   y   y   y   n   n   n   y   n   y   y   y   n   y  
414 republican y   y   y   y   y   y   y   y   n   y   ?   ?   ?   y   n   y  
416 democrat   n   y   y   n   n   y   y   y   ?   y   n   n   n   n   n   y  
417 republican y   y   n   y   y   y   n   n   n   y   n   n   y   y   n   y  
418 democrat   y   y   y   n   n   n   y   y   y   y  

We use the rest of the model data to test our model;consequently, it is assigned to testdata 
variable. 

In [47]:
testdata<-X[-test,]
nrow(testdata)

[1] 305

Logistic regression model is formed by using glm function in R language. The meaning of glm is fitting generalized linear models.

In [48]:
logitmodel<- glm(V1 ~ V4+V7+V9+V12+V15+V17, family=binomial(link="logit"), data= modeldata)
logitmodel

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Call:  glm(formula = V1 ~ V4 + V7 + V9 + V12 + V15 + V17, family = binomial(link = "logit"), 
    data = modeldata)

Coefficients:
(Intercept)          V4n          V4y          V7n          V7y          V9n  
   -33.4921      16.4759      13.8872      -0.6359      -1.7999      16.7255  
        V9y         V12n         V12y         V15n         V15y         V17n  
    15.6025       2.3715      -0.4687     -18.9774       0.6509       3.3996  
       V17y  
     1.7594  

Degrees of Freedom: 129 Total (i.e. Null);  117 Residual
Null Deviance:	    160.5 
Residual Deviance: 52.22 	AIC: 78.22

After the model is generated, we use summary code to get a short information about the model 
that we generate.

In [49]:
summary(logitmodel)
anova(logitmodel, test="Chisq")


Call:
glm(formula = V1 ~ V4 + V7 + V9 + V12 + V15 + V17, family = binomial(link = "logit"), 
    data = modeldata)

Deviance Residuals: 
     Min        1Q    Median        3Q       Max  
-2.34812  -0.12534  -0.00003   0.19392   2.35029  

Coefficients:
              Estimate Std. Error z value Pr(>|z|)  
(Intercept)   -33.4921 23196.2466  -0.001   0.9988  
V4n            16.4759  8260.9170   0.002   0.9984  
V4y            13.8872  8260.9170   0.002   0.9987  
V7n            -0.6359 17872.5265   0.000   1.0000  
V7y            -1.7999 17872.5266   0.000   0.9999  
V9n            16.7255 12263.6011   0.001   0.9989  
V9y            15.6025 12263.6012   0.001   0.9990  
V12n            2.3715     3.6601   0.648   0.5170  
V12y           -0.4687     3.6402  -0.129   0.8975  
V15n          -18.9774  2249.7087  -0.008   0.9933  
V15y            0.6509     1.5986   0.407   0.6839  
V17n            3.3996     1.6284   2.088   0.0368 *
V17y            1.7594     1.0397   1.692   0.0906 .
---

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"

Df Deviance  Resid. Df Resid. Dev Pr(>Chi)    
NULL NA        NA 129       160.48286            NA
V4    2 74.683736 127        85.79912  6.062237e-17
V7    2  2.081378 125        83.71775  3.532113e-01
V9    2  4.158100 123        79.55965  1.250490e-01
V12   2 10.652829 121        68.90682  4.861470e-03
V15   2 10.890242 119        58.01658  4.317317e-03
V17   2  5.793037 117        52.22354  5.521511e-02

After the model is formed, we use predict function to estimate the model for testdata variable 
which is the rest of the modeldata.

In [50]:
prelogitmodel<-predict(logitmodel,testdata, type = "response")

Now, `prelogitmodel` contains the probability that given MP is a republican or a democrat.  We need to set a threshold for the probability. If it is greater than 0.6then we will predict that MP is a republican

In [51]:
a<-ifelse(prelogitmodel>0.6,"republican","democrat")

In [52]:
table(a)

a
  democrat republican 
       207         98 

To compare logitmodel that we constitute and prelogitmodel that predict our model, we use
table function. We need to compare the number of democrat congressmen and republican congressmen.For this reason, we take V1 column of the testdata that consists of the class of the congressmen which are democrat and republican. Also, the table(a) that is formed by using prelogitmodel consists of the number of the congressmen that belongs to two class.

In [53]:
logistic_table1<-table(real=testdata$V1,predicted=a)

In [54]:
logistic_table1

            predicted
real         democrat republican
  democrat        167         10
  republican       40         88

When we look at the anova function that gives informations about prelogitmodel, we take out laws
that their residual deviance has a small differences. Therefore, V9 and V17 are taken out from 
the code to change our model.

In [55]:
logitmodel<- glm(V1 ~ V4+V12+V15, family=binomial(link="logit"), data= modeldata)
logitmodel

Warning message:
"glm.fit: fitted probabilities numerically 0 or 1 occurred"


Call:  glm(formula = V1 ~ V4 + V12 + V15, family = binomial(link = "logit"), 
    data = modeldata)

Coefficients:
(Intercept)          V4n          V4y         V12n         V12y         V15n  
    -20.283       20.359       17.420        1.059       -1.257      -18.065  
       V15y  
      1.216  

Degrees of Freedom: 129 Total (i.e. Null);  123 Residual
Null Deviance:	    160.5 
Residual Deviance: 61.38 	AIC: 75.38

In [56]:
summary(logitmodel)
anova(logitmodel, test="Chisq")


Call:
glm(formula = V1 ~ V4 + V12 + V15, family = binomial(link = "logit"), 
    data = modeldata)

Deviance Residuals: 
     Min        1Q    Median        3Q       Max  
-2.20985  -0.32660  -0.00007   0.42669   2.43218  

Coefficients:
            Estimate Std. Error z value Pr(>|z|)
(Intercept)  -20.283   8380.499  -0.002    0.998
V4n           20.359   8380.499   0.002    0.998
V4y           17.420   8380.499   0.002    0.998
V12n           1.059      1.731   0.612    0.541
V12y          -1.257      1.795  -0.700    0.484
V15n         -18.065   2327.614  -0.008    0.994
V15y           1.216      1.620   0.750    0.453

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 160.483  on 129  degrees of freedom
Residual deviance:  61.381  on 123  degrees of freedom
AIC: 75.381

Number of Fisher Scoring iterations: 19


Df Deviance  Resid. Df Resid. Dev Pr(>Chi)    
NULL NA        NA 129       160.48286            NA
V4    2 74.683736 127        85.79912  6.062237e-17
V12   2  9.751598 125        76.04753  7.628997e-03
V15   2 14.666445 123        61.38108  6.534643e-04

When we look at the table above, the differences between the residual deviances are close to each other. Consequently, we constitute our logitmodel again without V7, V9 and V17 columns.

In [57]:
prelogitmodel<-predict(logitmodel,testdata, type = "response")

In [75]:
a<-ifelse(prelogitmodel>0.6,"republican","democrat")

In [76]:
table(a)

a
  democrat republican 
       206         99 

In [77]:
logistic_table2<-table(real=testdata$V1,predicted=a)

In [78]:
logistic_table2

            predicted
real         democrat republican
  democrat        172          5
  republican       34         94

According to table functions, the total number of congressmen that is known wrongly is 49; therefore, these two table functions are compared by using Pearson's Chi-squared test.
To use Pearson's Chi-squared test, chisq.test(matrix()) is used and compared it X-squared values. The logitmodel is statisticaly good when it has a big X-squared value. 

In [62]:
chisq.test(logistic_table1)
chisq.test(logistic_table2)


	Pearson's Chi-squared test with Yates' continuity correction

data:  logistic_table1
X-squared = 132.75, df = 1, p-value < 2.2e-16



	Pearson's Chi-squared test with Yates' continuity correction

data:  logistic_table2
X-squared = 165.74, df = 1, p-value < 2.2e-16


The second modal that is used has bigger X-squared value and it has less independent variables;consequently, using this model is  statistically good.

## 2.SVM Model
### a. Definition
 
 SVM is the contraction of Support Vector Machines.Cortes & Vapnik developed SVM's for binary classification. In this model, support vectors are determined by using the closest points that are the elements of the classes. Support vectors pass on the closest points and the distance between the two of support vectors are called margin.Also margin is maximized in this model. There is a hyperplane that seperates the support vectors and margin in the middle. In this model, each of classes shows a tendency to -1 or 1. Therefore, there are two hyperplanes that are $w.x_i>=  1$  when $y_i=1$ and $w.x_i <=  -1$ when $y_i=-1$ for $x_i$'s are the set of input, $y_i$ is set of output corresponding to $x_i$ and $w$ is the weight vector that predicts the $yi$ value. $H_1$ and $H_2$  are planes such that  $H_1:w.x_i=1$ and $H_2:w.x_i= -1$. Also there is a plane in the middle of this planes such that $H_0: w.x_i=0$. The shortest distance between $H_0$ and $H_1$ is called $+d$ and the shortest distance between $H_0$ and $H_2$ is called $-d$.
### b.Equations of SVM Model
Considering the movement of the support vector changes the boundary; concequently, the form of equation of hyperplanes occurs such that $w^T.x+b=0$ where $w$ is a weigth vector, $x$ is an input vector and $b$ is bias.
The distance between $H_0$ and $H_1$ is $(w.x+b)/||w||= 1/||w||$, then the margin is $2/||w||$. To maximize margin,$w$ can be minimized. When hyperplanes that we mentioned are gathered up, the equation $y_i(w.x_i)>=1$ is formed.
Suppose that we have a problem and we need to minimize $w$, $$minf(x)=(1/2)||w||^2$$ and $g(x)=y_i(w.x_i)+b=1$ or $g(x)$ can be written as $g(x)=(y_i(w.x_i)+b)-1=0$. It can be solved by Langrange Multiplier and we have two constraints that one of them is $g(x)=0$ and the solution maximum or minimum according to the gradient of $f or g$. 
When these two constraints are considered the Langrange Equation must be like this
$L(x,a)= f(x)- ag(x)$ and the derivative of $L(x,a)$ is zero. In general $$L(x,a)= f(x)- \sum_i a_i g_i$$. 
In this case given f(x) and g(x) we have Langrangian 
$$min Langrangian= (1/2)||w||^2 - \sum_i a_i((y_i(w.x_i)+b)-1)$$ 
 and the last form of the equation is
 $$min Langrangian= (1/2)||w||^2 - \sum_i a_i((y_i(w.x_i+b)+\sum_i a_i$$. 
 
   In SVM model the Langrangian is $$(1/2)||w||^2 - \sum_{i=1}^l a_i((y_i(w.x_i+b)+\sum_{i=1}^l a_i$$ where $i,a_i>0$ and $l$ is the number of training points. The derivatives of Langrangian with respect to $w$ and $b$ are zero therefore, we get 
 $$w=\sum_{i=1}^l a_iy_ix_i$$ and $$\sum_{i=1}^l a_iy_i=0$$. When we rewrite the minimum Langrangian by putting $w$ and  $b=\sum_{i=1}^l a_iy_i=0$ then we have maximum Langrangian
 $$max L=\sum_{i=1}^l a_i- (1/2)\sum_{i=1}^l a_ia_jy_iy_j(x_i.x_j)$$ where $a_i$'s are support vectors and positive.
 In this equation we learn the margin is maximized or not by using the inner product of $(x_i$ and $x_j$ and owing to Kernel function, this computation is easier because Kernel function $$K(x_i,x_j)=\phi(x_i).\phi(x_j)$$ describes the inner product or resembels in transformed space.

### SVM Kernels

1.Polinomial Kernel
 
 $$K(x,y)=(x.y+1)^p$$
 
 2.Radial Basis Kernel  
 
 $$K(x,y)=exp{-(|x-y|^2)/2}\sigma^2$$
 
 3.Sigmoid Kernel
 
 $$K(x,y)=tanh(\kappa x.y-\delta)$$
 


First of all, we need to download the library of SVM for using the model in R. Therefore, we use this code below

In [63]:
library (e1071)

After we downloaded the library, we construct our SVM model by using modeldata and classes that are the first column of this data. 

In [64]:
svmmodel<- svm(V1 ~ . , data=modeldata)
svmmodel


Call:
svm(formula = V1 ~ ., data = modeldata)


Parameters:
   SVM-Type:  C-classification 
 SVM-Kernel:  radial 
       cost:  1 
      gamma:  0.03030303 

Number of Support Vectors:  45


After the svmmodel is constructed, to take a short information about svmmodel, summary() function can be used as the following.

In [65]:
summary(svmmodel)


Call:
svm(formula = V1 ~ ., data = modeldata)


Parameters:
   SVM-Type:  C-classification 
 SVM-Kernel:  radial 
       cost:  1 
      gamma:  0.03030303 

Number of Support Vectors:  45

 ( 22 23 )


Number of Classes:  2 

Levels: 
 democrat republican




As indicated above the C-classification is one of the dual representation of SVM and in this type of classification the error function minimized. The minimized error function is $(1/2)w w^T + C\sum{i=1}^n\xi_i$ subject to constraints: $y_i(w^T\phi(x_i)+b)>=1-\xi_i$ where  $\xi_i>0$ and $i=1,2,...,N$. In error function $C$ is the capasity constant, $w$ is weight vector or vector of coefficient, $b$ is constant, $i$ is the number of training, $y_i$ is the sign of class and $x_i$ is the set of independent variables. Besides the kernel $\phi(x_i)$ converts from the input data(independent variables) to feature space.

We need to predict that the svmmodel after we set up model by using testdata. Also predict()
function is used to predict svmmodel and it assigns svmpredict variable.

In [66]:
svmpredict<- predict(svmmodel,testdata)
svmpredict

1          2          3          5          9         10         12 
republican republican   democrat   democrat republican   democrat republican 
        13         14         15         17         19         20         21 
  democrat   democrat republican   democrat republican   democrat   democrat 
        23         24         26         27         29         30         31 
  democrat   democrat   democrat   democrat republican   democrat republican 
        33         34         35         36         37         39         40 
  democrat republican   democrat republican republican republican   democrat 
        41         42         45         48         49         50         52 
  democrat   democrat   democrat   democrat   democrat republican republican 
        54         55         56         58         61         62         63 
republican   democrat republican republican   democrat republican   democrat 
        64         66         67         68         70         71         73 
  democrat republican republican republican   democrat   democrat   democrat 
        74         75         79         80         81         83         84 
republican   democrat   democrat republican   democrat republican republican 
        85         86         87         89         90         91         92 
republican   democrat republican   democrat republican   democrat   democrat 
        94         97         99        100        102        103        104 
  democrat   democrat   democrat republican   democrat   democrat   democrat 
       107        108        109        110        111        112        113 
republican   democrat   democrat   democrat   democrat republican   democrat 
       114        116        117        118        120        121        123 
republican   democrat   democrat republican republican republican republican 
       126        129        131        133        134        135        136 
republican   democrat   democrat republican republican republican republican 
       137        139        141        143        144        146        147 
republican   democrat republican republican   democrat   democrat republican 
       148        149        150        152        153        155        156 
  democrat republican   democrat republican   democrat republican republican 
       158        160        161        163        165        166        167 
  democrat   democrat republican   democrat   democrat   democrat republican 
       168        169        170        172        173        175        177 
  democrat republican   democrat republican   democrat   democrat   democrat 
       178        179        182        184        185        186        187 
  democrat   democrat   democrat   democrat   democrat   democrat   democrat 
       188        189        190        191        192        193        194 
  democrat republican   democrat republican republican   democrat   democrat 
       195        198        200        201        203        204        206 
  democrat republican   democrat   democrat   democrat   democrat   democrat 
       207        208        209        210        211        214        215 
republican republican   democrat   democrat   democrat   democrat republican 
       216        218        219        220        221        222        226 
republican republican   democrat   democrat   democrat   democrat republican 
       227        228        229        231        232        233        234 
  democrat   democrat republican republican republican   democrat republican 
       235        237        238        239        240        241        242 
  democrat   democrat   democrat   democrat republican republican   democrat 
       243        245        246        248        249        250        251 
  democrat   democrat   democrat republican   democrat   democrat republican 
       252        253        254        255        256        257        258 
republican   democrat republica

In [67]:
summary(svmpredict)

democrat republican 
       181        124

As noted above summary() code gives a short brief summary about svmpredict variable.

Finally, we have to compare between the real data and the model that we set up by using SVM. 
When this model is compared with real data, the only column that needed to be taken is the 
first column because it has the classes of congressmen that are democrat and republican.
Besides, we use table()code to make comparison between the real data and svmmodel and we assign the SVM_table variable because we will not write the table data later.

In [68]:
 SVM_table<-table(testdata$V1,svmpredict)

In [69]:
SVM_table

            svmpredict
             democrat republican
  democrat        170          7
  republican       11        117

## ANALYSIS
### Logistic Regression Model Process
When we formed the first logistic regression  model we used $$glm(V1~.,family=binomial(link="logit"), data= modeldata)$$code but it did not converge.For this reason we examined the data set and we tried to figure out the relationship between the class of congressmen and statues. Besides, we swithced the code to $ glm(V1 ~ V4+V7+V9+V12+V15+V17, family=binomial(link="logit"), data= modeldata)$ then this code converges but still it is not a best form of our logistic regression model because we have many independent variables in this code. Therefore, we use $anova(logitmodel,test='Chisq')$ code to state which of columns are used or not. When the colums are determined, the table of anova function is researched. The residual deviance of anova table tells the dependency of the columns and the class of congressmen by using the difference between the residual deviance for  each of columns. When the difference of residual deviance is greater than the other one's, this means that the column has a stronger relationship than the others. Hence these columns are choosed by using method that is mentioned.
Later we constituted the second logistic regression model by using the columns that are choosed again. After the prediction function of logitmodel formed, constructing a table for logitmodel we took a threshold for the probability. Then we got a table for comparing the real data.
Besides we compared which model is better by using Chi-Square method and the second model is determined. The reason of this detection is having less independent variables although the second logistic regression model has a small chi-square number for some cases.
### SVM Model Process
To begin with we download a library for using SVM model. Then we constructed our SVM model simply because it did not get a warning for not converging. Moreover, we predicted the SVM model that we formed and we made a comparison between the prediction of this model and the real data by using table function. According to table data,there are 18 congressmen are known wrongly which means that there are congressmen that are democrat but they pretend republican or vise versa. It is significant because if the republican or democrat congressmen attract supporter for themselves then they will need to find this off diagonal terms. 
### Chi-Square Method
Chi-Square method is one of the tests that detemines the existence of the relationship between two variables that are numerical or not in statistic. This method makes a comparison between the observed values and the expected values theoritically. To begin with, there is a subtraction between observed values and expected values. Then this test squares the difference that it found and divided by expected values. The formula of Chi-Square is as the following
$$\chi^{2}_c=\sum_{i=1}^{k}\frac{(O_i-E_i)^2}{E_i}$$

where the $O_i$ is the observed values, $E_i$ is the expected values and $c$ is the degree of freedom.
Chi- Square method tests whether there is a dependency in data set. Also if the calculated value of this method is bigger than the value of table data then there will be a relationship between the variables.
### Chi-Square Method in R
When we use this method we use chisq.test(matrix()) code and we determine which model is statistically good according to the results of the chisq.test()code.


In [79]:
chisq.test(logistic_table2)
chisq.test(SVM_table)


	Pearson's Chi-squared test with Yates' continuity correction

data:  logistic_table2
X-squared = 165.74, df = 1, p-value < 2.2e-16



	Pearson's Chi-squared test with Yates' continuity correction

data:  SVM_table
X-squared = 231.85, df = 1, p-value < 2.2e-16


According to the results of these codes, if we compare of two model that we constituted the SVM model is better than the Logistic Regression Model because the X-squared value of SVM model is greater than Logistic Regression model and  the SVM model is easier to form. 

## References
[1]C. Manning.(2007),Logistic regression,p1.
 [2]R.B. Burns & R. Burns, Business Research Methods and Statistics using SPSS(2008),p.573
 [3]R.Berwick & V.Idiot,(n.d.) An Idiot’s guide to Support vector machines (SVMs),pp.4-10.
 [4]Support Vector Machines(SVM),Retrieved from http://www.statsoft.com/Textbook/Support-Vector-Machines

 [5]S.Deviant,The Practically Cheating Statistics Handbook,Retrieved from http://www.statisticshowto.com/probability-and-statistics/chi-square/